### DISCUSSION TITLE
How to Embed WebView2 Runtimes in .NET Framework Class Library With Fody Weavers

### DISCUSSION THREAD

**Discussion by protegesolutions** [MicrosoftEdge/WebView2Feedback#3970] [Q&A]
I've read the two or three (closed) posts that mention the need to do this but they don't seem to resolve the issue.

We have a class library (consumed by third-party apps not under our control, and obviously a single file with embedded references) that previously used the _WebBrowser_ control. That was a very simple, easy to use and deploy control that is now past its prime. But trying to employ _WebView2_ has been quite a struggle. Now we have it working but we can't get the runtimes to be included in the dll. (We would have settled for the _WebView_ control but that has issues running under Admin rights, bizarrely.)

The library is .NET 4.6.2 (for as much backward compatibility in the .NET Framework as possible). We use **Fody Weavers** to include _RestSharp_, _NLog_, etc and have now added _Microsoft.Web.WebView2.Core_, _Microsoft.Web.WebView2.WinForms_ and _Microsoft.Web.WebView2.Wpf_.

We also have the Costura directive _UseRuntimeReferencePaths='true'_ which I believe is required for .NET Framework projects.

But when I run the app test form - which consumes the library - I get a path error:

    The path is not of a legal form.
    at System.IO.Path.NormalizePath(String path, Boolean fullCheck, Int32 maxPathLength, Boolean expandShortPaths)
    at System.IO.Path.InternalGetDirectoryName(String path)
    at Microsoft.Web.WebView2.Core.CoreWebView2Environment.LoadWebView2LoaderDll()
    at Microsoft.Web.WebView2.Core.CoreWebView2Environment.<CreateAsync>d__3.MoveNext()
    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
    at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
    at Microsoft.Web.WebView2.WinForms.WebView2.<InitCoreWebView2Async>d__23.MoveNext()

I've tried a few things to ensure that the runtimes are included such as embedding them as resources, to no avail.

This is a deal breaker for us if we can't get this working; we MUST have everything included in a single library/dll. So my question is, has anyone managed to do this? (Praying that the answer is, "Sure! Just do this ...".

**Comment by yunate** [comment]
Hi, @protegesolutions, thanks for your feedback!
The runtime of weview2 in .NET are （let's call it |runtime DLL|）:

- Microsoft.Web.WebView2.Core.dll
- Microsoft.Web.WebView2.WinForms.dll // optional WinForms application need this
- Microsoft.Web.WebView2.Wpf.dll // optional WPF application need this
- WebView2Loader.dll

Could you help to check if |runtime DLL| all exist?

And about:
" _I've tried a few things to ensure that the runtimes are included such as embedding them as resources; we MUST have everything included in a single library/dll_". 
Do you mean: You want to package |runtime DLL| into a resource, and when the application run, release them again?

This may be helpful: [https://github.com/yunate/Temp](https://github.com/yunate/Temp)

**Comment by protegesolutions** [comment]
Hi @yunate. Thanks for your reply.

Please also refer to https://github.com/MicrosoftEdge/WebView2Feedback/issues/2645 where I've added a bit more about our situation.

In the original question here, I'm pretty sure, tho not 100% certain, that it is the _WebView2Loader.dll_ references which are not being "found". (I referred to "runtimes" because the files are located in a folder called _runtimes_, emitted by the compiler.)

As I mention in the question, we were hoping to replace _WebBrowser_ with _WebView2_ but the deployment (and other*) complications of the latter are making me nervous about fragility, portability, etc.

\* We spent a couple of days receiving _InvalidCastExceptions_ which we eventually tracked down to problems with the _.csprog_ files, but that's another story. We just want something simple and bulletproof to handle a pop-up _oAuth2_ form within the library but I'm beginning to think that we simply have to move that functionality out of the library, which I can see could make more design sense, though it makes things more complicated for the consumer of the library.

**Comment by yunate** [comment]
- Microsoft.Web.WebView2.Core.dll
- Microsoft.Web.WebView2.WinForms.dll // optional WinForms application need this
- Microsoft.Web.WebView2.Wpf.dll // optional WPF application need this
- WebView2Loader.dll

It would be better for the dynamic library above to be seem as dependency. This are used in compilation and should be packaged into application's installer.
 
The runtime that you confused are msedgewebview2.exe. EmbeddedBrowserWebView.dll and some other files (the dir is: `C:\Program Files (x86)\Microsoft\EdgeWebView\Application\103.0.1264.77`). This must be installed on the target machines.
This is so-called runtime.
![image](https://user-images.githubusercontent.com/37793342/182753375-585bbcb9-c524-432d-b685-fa90e6c06557.png)
